src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.Cache.cs

```diff
@@ -446,84 +446,47 @@ private static bool IsParameterRefEqual(in ParameterRef parameterRef, ReadOnlySp
         /// </summary>
         // AggressiveInlining used since this method is only called from two locations and is on a hot path.
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        public static ulong GetKey(ReadOnlySpan<byte> propertyName)
+        public static ulong GetKey(ReadOnlySpan<byte> name)
         {
-            const int BitsInByte = 8;
             ulong key;
-            int length = propertyName.Length;
+
+            ref byte reference = ref MemoryMarshal.GetReference(name);
+            int length = name.Length;
 
             if (length > 7)
             {
-                key = MemoryMarshal.Read<ulong>(propertyName)
-                    & 0x00FFFFFFFFFFFFFF
-                    // Include the length with a max of 0xFF.
-                    | ((ulong)Math.Min(length, 0xFF)) << (7 * BitsInByte);
-            }
-            else if (length > 3)
-            {
-                key = MemoryMarshal.Read<uint>(propertyName);
-
-                if (length == 7)
-                {
-                    key |= (ulong)propertyName[6] << (6 * BitsInByte)
-                        | (ulong)propertyName[5] << (5 * BitsInByte)
-                        | (ulong)propertyName[4] << (4 * BitsInByte)
-                        | (ulong)7 << (7 * BitsInByte);
-                }
-                else if (length == 6)
-                {
-                    key |= (ulong)propertyName[5] << (5 * BitsInByte)
-                        | (ulong)propertyName[4] << (4 * BitsInByte)
-                        | (ulong)6 << (7 * BitsInByte);
-                }
-                else if (length == 5)
-                {
-                    key |= (ulong)propertyName[4] << (4 * BitsInByte)
-                        | (ulong)5 << (7 * BitsInByte);
-                }
-                else
-                {
-                    key |= (ulong)4 << (7 * BitsInByte);
-                }
+                key = Unsafe.ReadUnaligned<ulong>(ref reference) & 0x00ffffffffffffffL;
+                key |= (ulong)Math.Min(length, 0xff) << 56;
             }
-            else if (length > 1)
+            else
             {
-                key = MemoryMarshal.Read<ushort>(propertyName);
+                key =
+                    length > 5 ? Unsafe.ReadUnaligned<uint>(ref reference) | (ulong)Unsafe.ReadUnaligned<ushort>(ref Unsafe.Add(ref reference, 4)) << 32 :
```

Comment:
Can we add a private six-byte struct to avoid the double read?

Reply:
Processors do not have 6-byte read instruction. Six-byte struct is still going to be JITed as double read, or maybe even as something less efficient.

Reply:
As @jkotas pointed out, there are no way to read 6 bytes using a single instruction. The current code is optimized well and uses only 4 instructions for that line: 

```assembly
mov       edx,[rax]            ; read first 4 bytes
movzx     r8d,word ptr [rax+4] ; read following 2 bytes
shl       r8,20                ; shift left two bytes by 32 bits
or        rdx,r8               ; put the result into a single register
```

Speaking about the regression, as you know the operating system easily can spoil results by doing something in background and this can be easily confirmed by running the benchmark several times.

|          Method |   value |     Mean |     Error |    StdDev |   Median | Code Size |
|---------------- |-------- |---------:|----------:|----------:|---------:|----------:|
| GetKeyOptimized | Byte[0] | 4.739 ns | 0.0039 ns | 0.0033 ns | 4.739 ns |     199 B |
| GetKeyOptimized | Byte[1] | 3.690 ns | 0.0033 ns | 0.0026 ns | 3.690 ns |     199 B |
| GetKeyOptimized | Byte[2] | 5.341 ns | 0.1355 ns | 0.1201 ns | 5.296 ns |     199 B |
| GetKeyOptimized | Byte[3] | 4.350 ns | 0.1242 ns | 0.1741 ns | 4.296 ns |     199 B |
| GetKeyOptimized | Byte[4] | 5.266 ns | 0.0028 ns | 0.0026 ns | 5.266 ns |     199 B |
| GetKeyOptimized | Byte[5] | 4.542 ns | 0.0766 ns | 0.0679 ns | 4.550 ns |     199 B |
| GetKeyOptimized | Byte[6] | 4.994 ns | 0.0009 ns | 0.0008 ns | 4.994 ns |     199 B |
| GetKeyOptimized | Byte[7] | 3.948 ns | 0.0015 ns | 0.0014 ns | 3.947 ns |     199 B |
| GetKeyOptimized | Byte[8] | 3.916 ns | 0.1181 ns | 0.1768 ns | 3.802 ns |     199 B |
| GetKeyOptimized | Byte[9] | 4.022 ns | 0.0787 ns | 0.0736 ns | 4.020 ns |     199 B |

|          Method |   value |     Mean |     Error |    StdDev | Code Size |
|---------------- |-------- |---------:|----------:|----------:|----------:|
| GetKeyOptimized | Byte[0] | 4.665 ns | 0.0337 ns | 0.0331 ns |     199 B |
| GetKeyOptimized | Byte[1] | 3.985 ns | 0.0439 ns | 0.0367 ns |     199 B |
| GetKeyOptimized | Byte[2] | 5.288 ns | 0.0105 ns | 0.0088 ns |     199 B |
| GetKeyOptimized | Byte[3] | 4.239 ns | 0.0099 ns | 0.0077 ns |     199 B |
| GetKeyOptimized | Byte[4] | 5.552 ns | 0.0029 ns | 0.0025 ns |     199 B |
| GetKeyOptimized | Byte[5] | 4.474 ns | 0.1023 ns | 0.0906 ns |     199 B |
| GetKeyOptimized | Byte[6] | 4.738 ns | 0.0021 ns | 0.0017 ns |     199 B |
| GetKeyOptimized | Byte[7] | 3.949 ns | 0.0040 ns | 0.0034 ns |     199 B |
| GetKeyOptimized | Byte[8] | 4.201 ns | 0.1170 ns | 0.1037 ns |     199 B |
| GetKeyOptimized | Byte[9] | 3.706 ns | 0.0595 ns | 0.0464 ns |     199 B |

The tests don't use `AggressiveInlining` because the JIT compiler understand that there is an array and generates code slightly differently, not so much, but differently. And since a span is used to pass a name to the `GetKey` method we should see how it could be compiled. Therefore, numbers you see actually include some overhead caused by the following instructions:

```assembly
; Expirements.Program.GetKeyOptimized(Byte[])
       sub       rsp,38
       xor       eax,eax
       mov       [rsp+28],rax
       test      rdx,rdx
       jne       short M00_L00
       xor       ecx,ecx
       xor       eax,eax
       jmp       short M00_L01
M00_L00:
       lea       rcx,[rdx+10]
       mov       eax,[rdx+8]
M00_L01:
       lea       rdx,[rsp+28]
       mov       [rdx],rcx
       mov       [rdx+8],eax
       lea       rcx,[rsp+28]
       call      Expirements.Program.<GetKeyOptimized>g__GetKey|1_0(System.ReadOnlySpan`1<Byte>)
       nop
       add       rsp,38
       ret
; Total bytes of code 56
```

This code is identical for all benchmarks and gives equal overhead for all tests. So the code works faster than you can see for the optimizes variant and the original one, but the actual impact can be only measured by running JSON serialization tests because methods are aggressively inlined and could look differently.